In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'/Users/jazmingutierrez/DEC2023 Instacart Basket Analysis'

In [3]:
df_prods=pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Creating DataFrame for mixtype data 

In [6]:
df_prods.shape

(49693, 5)

In [5]:
# checking for data that would need to be investigated
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The Data contains incosistency of prices there is a max of 99999. It would be irrational for a max to be 99999 becuase the mean is 9.994.

In [6]:
# Searching for the price value 99999
df_prods[df_prods['prices'] ==99999.000000]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

In [7]:
# Searching for milk price
df_prods[df_prods['product_name'] == 'Milk']

product_id product_name  aisle_id  department_id  prices
16611       16611         Milk        84             16     7.3

In [29]:
# Replacing with new value
df_prods_new= df_prods.replace(to_replace= 99999.0, value = 7.3)

In [59]:
# Checking max output
df_prods_new.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.045445     67.770249      11.728433      7.682253
std    14344.080758     38.316774       5.850282      4.200158
min        1.000000      1.000000       1.000000      1.000000
25%    12422.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000     25.000000

Data still contains inconsistency of 14900 max and will be searched and replaced with proper price. 

In [31]:
# Searching for max price
df_prods_new[df_prods_new['prices'] == 14900.000000]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   

        prices  
21554  14900.0

In [32]:
# Replacing with new value
df_prods_new= df_prods_new.replace(to_replace= 14900.0, value = 7.3)

In [33]:
# Checking max output
df_prods_new.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.045445     67.770249      11.728433      7.682253
std    14344.080758     38.316774       5.850282      4.200158
min        1.000000      1.000000       1.000000      1.000000
25%    12422.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000     25.000000

In [34]:
# Create DataFrame
df_test = pd.DataFrame()

In [35]:
# Create a mix type column 
df_test['mix'] = ['a', 'b', 1, True]

In [36]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [37]:
# Check for mix types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0: 
        print (col)

mix


In [38]:
df_test['mix'] = df_test['mix'].astype('str')

# Fix missing Values

In [39]:
# Find the count of missing observations (missing observation= count of 1's false=0)
df_prods_new.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

## Creating subset of new dataframe to view missing values

In [40]:
df_nan = df_prods_new[df_prods_new['product_name'].isnull() ==True]

In [41]:
df_nan

product_id product_name  aisle_id  department_id  prices
33           34.0          NaN       121             14    12.2
68           69.0          NaN        26              7    11.8
115         116.0          NaN        93              3    10.8
261         262.0          NaN       110             13    12.1
525         525.0          NaN       109             11     1.2
1511       1511.0          NaN        84             16    14.3
1780       1780.0          NaN       126             11    12.3
2240       2240.0          NaN        52              1    14.2
2586       2586.0          NaN       104             13    12.4
3159       3159.0          NaN       126             11    13.1
3230       3230.0          NaN       120             16    14.4
3736       3736.0          NaN        41              8    14.8
4283       4283.0          NaN        77              7    14.4
4790       4790.0          NaN        91             16    14.5
38187     38183.0          NaN        39             12    20.9
40444     40440.0          NaN       120             16    14.8

## Addressing missing values

In [42]:
df_prods_new.shape

(49693, 5)

In [43]:
# Creating new dataframe for non missing values
df_prods_clean = df_prods_new[df_prods_new['product_name'].isnull() ==False]

In [44]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [45]:
# finding duplicates for df_prods_clean dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [46]:
df_dups

product_id                                       product_name  \
462         462.0                  Fiber 4g Gummy Dietary Supplement   
18459     18458.0                                         Ranger IPA   
26810     26808.0               Black House Coffee Roasty Stout Beer   
35309     35306.0  Gluten Free Organic Peanut Butter & Chocolate ...   
35495     35491.0                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# Addressing duplicates by dropping duplicates

In [47]:
df_prods_clean.shape

(49677, 5)

In [48]:
#create dataframe does not include duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [49]:
df_prods_clean_no_dups.shape

(49672, 5)

In [50]:
# Exporting dataframe
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5 Data Consistency on df_ords


# 2. Descriptive analysis on df_ords

In [51]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week   time_of_day  days_since_prior_order  
count        3.421083e+06  3.421083e+06            3.214874e+06  
mean         2.776219e+00  1.345202e+01            1.111484e+01  
std          2.046829e+00  4.226088e+00            9.206737e+00  
min          0.000000e+00  0.000000e+00            0.000000e+00  
25%          1.000000e+00  1.000000e+01            4.000000e+00  
50%          3.000000e+00  1.300000e+01            7.000000e+00  
75%          5.000000e+00  1.600000e+01            1.500000e+01  
max          6.000000e+00  2.300000e+01            3.000000e+01

The data looks to be in order the min and max have been analayzed for each column there are no numbers that need to investigated. 

# 3. Checking the mix type data

In [52]:
# Check for mix types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0: 
        print (col)

# 4. No mix type data

# 5. Checking for missing values in df_ords

In [53]:
# mssing observations are considered True 1 and False 0. summing up the total number of trues. 
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
time_of_day                    0
days_since_prior_order    206209
dtype: int64

There are total of 206,209 missing values for days_since_prior_order this could be because many users that order are new customers and have no prior days since order. 

In [54]:
# Creating a dataframe to view missing values 
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() ==True]

In [55]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         time_of_day  days_since_prior_order  
0                  8                     NaN  
11                11                     NaN  
26                14                     NaN  
39                11                     NaN  
45                12                     NaN  
...              ...                     ...  
3420930           12                     NaN  
3420934           18                     NaN  
3421002           19                     NaN  
3421019           15                     NaN  
3421069           11                     NaN  

[206209 rows x 7 columns]

# 6. Addressing the missing values

There are a signifcant amount of missing values for the column days_since_prior_order due to the fact that most users are new and cause Values to be counted as NaN. It would be irrational to to delete or replace data becuase of the large amount of missing vlaues 206209. 

# 7. Checking for duplicate values

In [56]:
#dataframe for duplicates in df_ords
df_ords_dups = df_ords[df_ords.duplicated()]

In [57]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, time_of_day, days_since_prior_order]
Index: []

There are no duplicates found in df_ords_dups.

# 8. Addressing the duplicates

No duplicates found. 

In [58]:
# Exporting data frame in prepared data as orders_checked.csv
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))